bugshaker/src/main/java/com/github/stkent/bugshaker/email/screenshot/maps/MapBitmapObservable.java

```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright 2016 Stuart Kent
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.github.stkent.bugshaker.email.screenshot.maps;
+
+import android.graphics.Bitmap;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.google.android.gms.maps.GoogleMap;
+import com.google.android.gms.maps.MapView;
+import com.google.android.gms.maps.OnMapReadyCallback;
+
+import rx.Observable;
+import rx.Subscriber;
+import rx.android.schedulers.AndroidSchedulers;
+
+final class MapBitmapObservable {
+
+    static Observable<LocatedBitmap> create(@NonNull final MapView mapView) {
+        final int[] locationOnScreen = new int[] {0, 0};
+        mapView.getLocationOnScreen(locationOnScreen);
+
+        final Observable<LocatedBitmap> result = Observable.create(new Observable.OnSubscribe<LocatedBitmap>() {
+            @Override
+            public void call(final Subscriber<? super LocatedBitmap> subscriber) {
+                mapView.getMapAsync(new OnMapReadyCallback() {
+                    @Override
+                    public void onMapReady(@NonNull final GoogleMap googleMap) {
+                        googleMap.snapshot(new GoogleMap.SnapshotReadyCallback() {
+                            @Override
+                            public void onSnapshotReady(@Nullable final Bitmap bitmap) {
+                                if (bitmap != null) {
+                                    subscriber.onNext(
+                                            new LocatedBitmap(bitmap, locationOnScreen));
+
+                                    subscriber.onCompleted();
+                                } else {
+                                    subscriber.onError(new NullPointerException());
+                                }
+                            }
+                        });
+                    }
+                });
+            }
+        });
+
+        // Required, since MapView.onMapReady _must_ be called from the main thread.
+        return result.subscribeOn(AndroidSchedulers.mainThread());
```

Comment:
Ok, I tested this as follows:

- removed the `subscribeOn` call from this class;
- ran the example app with the following configurations at the subscription site:

1
===
```java
screenshotProvider.getScreenshotUri(activity)
                            .subscribeOn(AndroidSchedulers.mainThread())
                            .subscribe(...)
```

This resulted in the app working as expected.

2
===
```java
screenshotProvider.getScreenshotUri(activity)
                            .subscribeOn(Schedulers.io())
                            .subscribe(...)
```

This resulted in the app crashing with the following stack trace:

```
java.lang.IllegalStateException: Fatal Exception thrown on Scheduler.Worker thread.
    at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:62)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:423)
    at java.util.concurrent.FutureTask.run(FutureTask.java:237)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:269)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
    at java.lang.Thread.run(Thread.java:818)
Caused by: rx.exceptions.OnErrorFailedException: Error occurred when trying to propagate error to Observer.onError
    at rx.observers.SafeSubscriber._onError(SafeSubscriber.java:194)
    at rx.observers.SafeSubscriber.onError(SafeSubscriber.java:120)
    at rx.internal.operators.OperatorSubscribeOn$1$1$1.onError(OperatorSubscribeOn.java:71)
```

3
===
```java
screenshotProvider.getScreenshotUri(activity)
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribeOn(Schedulers.io())
                            .subscribe(...)
```

This resulted in the app crashing with the following stack trace:

```
java.lang.IllegalStateException: getMapAsync() must be called on the main thread
    at com.google.android.gms.common.internal.zzx.zzcD(Unknown Source)
    at com.google.android.gms.maps.MapView.getMapAsync(Unknown Source)
    at com.github.stkent.bugshaker.email.screenshot.maps.MapBitmapObservable$1.call(MapBitmapObservable.java:42)
    at com.github.stkent.bugshaker.email.screenshot.maps.MapBitmapObservable$1.call(MapBitmapObservable.java:37)
    at rx.Observable.unsafeSubscribe(Observable.java:8098)
    ...
```

4
===
```java
screenshotProvider.getScreenshotUri(activity)
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribeOn(Schedulers.io())
                            .subscribeOn(AndroidSchedulers.mainThread())
                            .subscribe(...)
```

This results in the same error as in case 3.

5
===
```java
screenshotProvider.getScreenshotUri(activity)
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribeOn(AndroidSchedulers.mainThread())
                            .subscribeOn(Schedulers.io())
                            .subscribe(...)
```

This resulted in the app working as expected.

===
I'll post reasoning about these cases in a separate comment.